pasha/context.py

```diff
@@ -0,0 +1,278 @@
+
+# Distributed under the terms of the BSD 3-Clause License.
+# The full license is in the file LICENSE, distributed with this software.
+#
+# Author: Philipp Schmidt <philipp.schmidt@xfel.eu>
+# Copyright (c) 2020, European X-Ray Free-Electron Laser Facility GmbH.
+# All rights reserved.
+
+import mmap
+from multiprocessing import get_context
+from multiprocessing.pool import ThreadPool
+from os import cpu_count
+from queue import Queue
+from threading import local
+
+import numpy as np
+
+from .functor import Functor
+
+
+class MapContext:
+    """Context to execute map operations.
+
+    A map operation applies a single callable to each item in a
+    collection. The context define the runtime conditions for this
+    operation, which may be in a process pool, for example.
+
+    As some of these environments may require special memory semantics,
+    the context also provides a method of allocating ndarrays. The only
+    abstract method required by an implementation is map. It is
+    recommended to use the staticmethod run_worker in this type for the
+    actual mapping loop in each worker.
+    """
+
+    def __init__(self, num_workers):
+        """Initialize this map context.
+
+        In order to use the default allocation methods in this type,
+        in particular array_per_worker(), the number of workers in this
+        context need to be passed to its initializer or stored in a
+        `num_workers` property.
+
+        Args:
+            num_workers (int): Number of workers used in this context.
+        """
+
+        self.num_workers = num_workers
+
+    def array(self, shape, dtype=np.float64):
+        """Allocate an array shared with all workers.
+
+        The implementation may decide how to back this memory, but it
+        is required that all workers of this context may read AND write
+        to this memory. The default implementation allocates directly
+        on the heap.
```

Comment:
Maybe call this the 'base' implementation instead of the 'default'? From the user's POV, the default implementation is the one provided by the default context, i.e. ProcessContext using shared memory. I think the `@wraps` decorators will copy this docstring onto the top-level functions, so it won't be obvious that it's talking about the base class.